### ISSUE TITLE
PPO training

### ISSUE THREAD

**Issue by mehdimashayekhi** [openai/lm-human-preferences#1]
Hi, thanks much for sharing. I have a quick question, when training policy using PPO, here,             https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L341), aren't you using same policy as to generate rollout (https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L285) ?; except for training you are dividing logits by Temperature. Maybe I am missing sth but for PPO,  to calculate loss we need to have pi(theta)/pi(old) as in ( https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L351), but it seems the old and new policy are the same, I really appreciate your answer

**Comment by WuTheFWasThat** [comment]
Hi,

The code is a bit confusing, but I believe it's correct.  

If you look here:
https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L180
Each call to this function `train`, is using the same rollouts - the `rollouts` object stores the logprobs which were generated from before any updates.  But each call to `train_minibatch` will update the network.  So after the first minibatch, `self.policy` will be different and produce different logprobs than what the values stored in the `rollouts` object are.  Does that make sense?

**Comment by mehdimashayekhi** [comment]
Thank you for your reply, but still probably I am missing something, when you call ```train``` and subsequently ```train_minibatch```, before updating the network (https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L176) ```loss``` function is called (https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L319) and inside this function, policy is not updated yet, and https://github.com/openai/lm-human-preferences/blob/f774e89b9e3762a03b7f3953189861979bee1775/lm_human_preferences/train_policy.py#L341 uses the same policy as rollouts ? thanks

**Comment by WuTheFWasThat** [comment]
That function is just constructing the tensorflow graph.  The op is not actually called until `train_minibatch` is called

**Comment by mehdimashayekhi** [comment]
ahhh, got it, I missed the part that ```train_minibatch``` is called multiple times, and as you said earlier, after the first minibatch, ```self.policy``` will be different and produce different ```logprobs``` than what the values stored in the ```rollouts``` object are. Thanks much again.

**Comment by WuTheFWasThat** [comment]
no worries!  it's definitely a bit confusing, sorry about that.  PRs to clean it up welcome :P
